azurerm/resource_arm_user_assigned_identity.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmUserAssignedIdentity() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmUserAssignedIdentityCreate,
+		Read:   resourceArmUserAssignedIdentityRead,
+		Update: resourceArmUserAssignedIdentityCreate,
+		Delete: resourceArmUserAssignedIdentityDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": resourceGroupNameSchema(),
+			"location":            locationSchema(),
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"principal_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmUserAssignedIdentityCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).userAssignedIdentitiesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for Azure ARM user identity creation.")
+
+	resourceName := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+	identity := msi.Identity{
+		Name:     &resourceName,
+		Location: &location,
+		Tags:     expandTags(tags),
+	}
+
+	identity, err := client.CreateOrUpdate(ctx, resGroup, resourceName, identity)
+	if err != nil {
+		return fmt.Errorf("Error Creating/Updating User Assigned Identity %q (Resource Group %q): %+v", resourceName, resGroup, err)
+	}
+
+	if identity.ID == nil {
+		return fmt.Errorf("Cannot read User Assigned Identity %q ID (resource group %q) ID", resourceName, resGroup)
+	}
+
+	d.SetId(*identity.ID)
+
+	return resourceArmUserAssignedIdentityRead(d, meta)
+}
+
+func resourceArmUserAssignedIdentityRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).userAssignedIdentitiesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	resourceName := id.Path["userAssignedIdentities"]
+
+	resp, err := client.Get(ctx, resGroup, resourceName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on User Assigned Identity %q (Resource Group %q): %+v", resourceName, resGroup, err)
+	}
+
+	if resp.IdentityProperties == nil || resp.IdentityProperties.PrincipalID == nil {
+		return fmt.Errorf("Error PrincipalID can't be null for User Assigned Identity %q (Resource Group %q): %+v", resourceName, resGroup, err)
+	}
+
+	d.Set("resource_group_name", resGroup)
+	d.Set("location", resp.Location)
+	d.Set("name", resp.Name)
+	d.Set("principal_id", resp.IdentityProperties.PrincipalID.String())
```

Comment:
PrincipalID is also a pointer, so we should check it to avoid null pointer crashes.. 
Would you recommend to keep what I have now or replace it with this?

```
if props := resp.IdentityProperties; props != nil {
  if principalId:= props.PrincipalId; principalId != nil {
    d.Set("principal_id", principalID.String())
  }
}
```

Reply:
I also think lack of PrincipalID should be treated as an error because it's used to pass to other resources.. So, if Identity was provisioned but for some reason didn't return PrincipalID code shouldn't silently ignore it